### PR TITLE
136: Transcripts closed captions

### LINKED ISSUE #131: Upgrade to Node.js 18+

### LINKED ISSUE #136: Display transcripts on landing page route
If the feed has a `<podcast:transcript>`, we should display the transcript in conjunction with the landing pages.

### LINKED ISSUE #137: Closed Captioning in embed player
If the feed has `<podcast:transcript>` available we should display the closed caption.

Questions:
- Closed Captioning when timings are wrong (not present) (link to #136?)
- How to handle large chunks of text in the timestamp, that would be too large for display in the fixed-height display?

### LINKED ISSUE #139: Persistent player on /listen defaults to latest episode, not the episode in the query params
When opening a url like this:

https://play.prx.org/listen?ge=prx_72_8bcdb9a8-6b72-4dce-8af9-5816ad9ef0f8&uf=http%3A%2F%2Ffeeds.radiotopia.fm%2Fradiotopia-showcase

image

the episode info is correctly loaded at the top section, but the persistent player is defaulted to the latest episode, so clicking on the play bottom at top plays the episode from the `ge=[guid]` param, but clicking the play button in the persistent player at the bottom plays the latest episode instead.

We should have the persistent player at bottom use the `ge` episode when provided in the initial request.

### PR DESCRIPTION
**Pull request by rpeterman-gp**
Closes #136 
Closes #137 
Closes #139 
Closes #131 

- Add podcast transcripts to RSS feed parser.
- Add API's to proxy and convert to VTT and JSON transcripts. 
- Add transcripts to episode details of listen pages.
- Add closed captions ticker to default embed player.
- Add closed caption feed to listen page and cover art embed player.

## To Review

- [x] Checkout Branch.
- [x] Run `asdf install`.
- [x] Run `yarn`.
- [x] Run `yarn dev`.

> ...then...

- [ ] Transcript without speakers denoted, low fidelity VTT
  - [ ] Embed: http://localhost:4300/e?uf=https%3A%2F%2Fkala.farski.com%2Fearhustlesq.xml
  - [ ] Cover Art Embed: http://localhost:4300/e?uf=https%3A%2F%2Fkala.farski.com%2Fearhustlesq.xml&ca=1
  - [x] Listen Page: http://localhost:4300/listen?uf=https%3A%2F%2Fkala.farski.com%2Fearhustlesq.xml
- [ ] Transcripts with speakers denoted, multiple transcripts, high fidelity JSON available
  - [ ] Embed: http://localhost:4300/e?uf=https%3A%2F%2Ffeeds.buzzsprout.com%2F231452.rss&ac=76B54D&ac=30A7CF&ac=f130c0&sp=20
  - [ ] Cover Art Embed: http://localhost:4300/e?uf=https%3A%2F%2Ffeeds.buzzsprout.com%2F231452.rss&ca=1&ac=76B54D&ac=30A7CF&ac=f130c0&sp=20
  - [x] Listen Page: http://localhost:4300/listen?uf=https%3A%2F%2Ffeeds.buzzsprout.com%2F231452.rss&ac=76B54D&ac=30A7CF&ac=f130c0&sp=20
    - [x] Click an episode title in the middle of the playlist.
    - [x] Reload the page and ensure that episode is what is displayed on the page and in the player.
 - [ ] Test links in Chrome.
 - [ ] Test links in Firefox.
 - [ ] Test links in Safari.
 - [ ] Test links in any other browser you have.

### CODE REVIEW

**Review by farski — CHANGES_REQUESTED** [review]
There seems to be an issue that is preventing the [CC] button from appearing consistently for the embed and cover art views.

**Review by brandonhundt — COMMENTED** [review]
@rpeterman-gp This works as expected after adding the API line to the `.env` file. So it is good to me. If @farski or @cavis could give you a solid code review, that'd be great. thanks!

**Review by farski — COMMENTED** [review]
A few things I spotted:

1. On the embed view, while you're scrubbing, the CCs don't update. They only update once you release the scrub. I think, if the transcript data is already loaded, having the text follow the play head would make it easy to find a particular spot in the episode. This is most practical on the embed view, since there's no competition with scrolling the text any other way.
1. On the cover art view and listen page, if you click the CC button but haven't hit play yet, there's no text visible (this may be related to No. 5 below, and have to do with the specific timing of captions. It's possible to reproduce with the Ear Hustle example.)
2. On the cover art view, if you have not manually scrolled the text at all, my expectation would be that the visible text always follows the playhead. Currently, if you scrub or use the skip buttons, the text will stay scrolled to whatever spot it's at, and you have to use the CC jump button to get them back in sync.
3. Conversely, if you manually scroll the text to a point ahead of the playhead the text stays locked to the point you scrolled to, but as the audio plays and the playhead catches up to the visible text, the text starts automatically following the playhead again. My expectation is that if I scroll ahead, and take no other actions, the text will stay scrolled to the spot I've picked, even if the playhead crosses that region. I think if I've scrolled the text away from the playhead, and then I use the scrubber or the skip buttons, I would expect those to control both the audio and the text. The current behavior leads to inconsistent results. For example, using the -5 Skip, usually results in the text and audio staying in sync, because the text for 5 seconds ago is still visible so it does the auto resync thing. But using the +30 Skip usually disconnects the audio and text, even if they had been in sync before hitting the button, because text 30s from now is generally off the screen. So I think any sort of scrub/skip action should resync the audio and text, and otherwise scrolling location should be maintained.
4. Sometimes, even during playback in the middle of an episode, when you enable the CC view for the CA or Listen pages, the text won't be visible. It seems to happen if you click the CC button in a gap between timestamped captions; until the playhead hits the next caption, no text will be displayed. I think it should always show the text if a transcript is available, even if there's no caption available for the specific moment the playhead is at. It could "round up" to the next available caption for deciding where in the text to make visible. (this may only happen with the lofi examples)
5. I find the word highlighting to be distracting and detrimental to readability for the high fidelity examples. Being able to compare the experience of the lofi example that doesn't do highlighting with the hifi example really drives home how much I prefer it without highlighting.
6. With the chat-style view, I find it strange that an individual speaker will move between the right and left sides of the view, depending on the sequencing of the different speakers. In the following example, Jordan is on the right, and then a couple sentences later Jordan is on the left. I think I would prefer if either: everyone were on the left, or each speaker is assigned a side and they are stuck to that side regardless of speaker sequencing. I would prefer having two "lefts" in a row than a speaker jumping between sides.

    image
7. On the CA and Listen view, when the text view first appears, it always shows the beginning of the transcript, and then jumps to the correct location. This causes an unnecessary flash before the view settles down.

**Comment by rpeterman-gp** [comment]
@farski Just push up some changes to address most of your notes.

Couldn't address note 1 do to the way the progress bar scrub state is handled. The scrub position is currently only scoped to the progress bar component state. Would need to refactor the root player state to include a scrub position, and could result in hammering a re-render of the entire app as you scrub, which could have performance problems. I will need more time to make that work smoothly. Also, having to find the correct caption cue out of thousands of cues as the user scrubs may not be performant. I am hoping the binary search I just added to the current caption render precess will make this feature possible.

I did my best to resolve note 8, but may still occur, though very briefly. Has to do with how long it takes to rendering hundreds to thousands of caption elements and their children. They have to all render before the current one can be scrolled to. I added a loader message to display when that render slow.

As for notes 6 and 7, those seem to be user preferences. How would we feel about adding a cc settings menu to control highlighting fidelity and caption layout?

**Review by farski — CHANGES_REQUESTED** [review]
1. Understood, may be a nice thing to work towards down the road.
2. This does seem to behave correctly now
3. This behaves better, but still not exactly how I would expect. Scrubbing or jumping to a time within some caption's range does work as expected, but scrubbing or jumping to a dead spot results in the text staying at the playhead's previous spot until playback hits the next caption. In some cases I hit this can mean waiting almost 10 seconds for the captions and playback to get back in sync.
4. The playhead moving through the point at which the captions were scrolled through no longer catches the captions and starts scrolling them unexpectedly. I still think any sort of scrub/skip action should resync the audio and text, even if the captions were manually scrolled, and that does not seem to be the case currently.
5. I do now always see captions when they are enabled, but the results don't always match my expectations. For example, with the EH clip (I'm looking at CA view), if I have CCs enabled and I put the playhead around 7:00 and start playback, when the playhead gets to ~7:07, it will be between two captions (i.e., no text is highlighted) and stay there until ~7:11 when the next caption starts. But if I pause playback, disable CCs, scrub directly to 7:07, and enable CCs, I always see the beginning of the captions until I hit play and the playhead reaches that next caption at ~7:11. I think it would be better if when the CCs become enabled, it always looks at the current playhead position, and scrolls the captions to the caption that matches that position ***or*** the next available caption in cases where there is no caption at that position. I think this should be true of any actions that cause the playhead to move also, which would resolve the related issue in No. 3.
6. And 7; I'm fine with making these sorts of things preferences if it doesn't over complicate the UI. If we are building these as accessibility features, I would like to make sure that our defaults are meeting the needs to of people relying on those features. My intuition is that for those users, for example, less highlighting is better, but I think we should engage with users/experts to help inform those sorts of choices.

**Comment by rpeterman-gp** [comment]
6. I should point out that the high fidelity highlighting is styled to not occur for users that have their `prefers-reduced-motion` set to `reduced`. You can test this out by opening the dev inspector and press Cmd+Shift+P and filtering for `reduced`. Select the "Emulate CSS prefers-reduced-motion: reduced" item. Repeat to turn off the emulation.

**Comment by rpeterman-gp** [comment]
Just pushed up a change that should address 3, 4, and 5, along with some performance optimizations.

**Comment by rpeterman-gp** [comment]
Pushed up changes to fix the build. Upgrade Node, Next, and React version, on long with several other dependencies. When reviewing, you may want to remove your `node_modules` directory, then run `yarn` before starting dev server.

**Review by farski — COMMENTED** [review]
The lo-fi sample seems to be in good shape; 3, 4, and 5 are working as I would expect.

I am seeing some strange behaviors with the hi-fi Buzzsprout example, though. I likely won't have time to dig in too much today and detail all the steps to reproduce, but just a couple quick ones I've noticed already:

If CCs are open when the page refreshes and the playhead is somewhere in the middle of the episode, after the page reloads if you open CCs before starting playback, it seems to remember where you were in the captions. But plackback always starts from the beginning. It was unexpected to see that state persist across page loads, but if it's going to I think the captions and playhead should behave the same way.

If you load the page, enable CCs, and then scrub or skip (before starting playback), the captions remain static and do not follow the playhead.

**Comment by rpeterman-gp** [comment]
Pushed up changes to update CC feed scroll with changes to player current time when audio has not been played. User had to have played the audio with an interaction before `timeupdate` events would trigger for scrub or skipped time changes.

As for the CC feed scroll position being remembered after a page reload, that is a Firefox bug (https://bugzilla.mozilla.org/show_bug.cgi?id=706792). However, I was able to find a way to handle it by conditionally reseting the scrollTop of the scroll area on init. Should still scroll to the current cue if you scrub or skip time before opening CCs.

**Review by brandonhundt — APPROVED** [review]
@rpeterman-gp This is very solid work. This is a strong place to be for this feature. Approving with the understanding that we will solicit user feedback, specifically around hi-fi transcripts and the styling in place there. :shipit:

File: components/Player/Player.tsx
Comment (cavis) [inline components/Player/Player.tsx:51]: Question: do we GET/load transcripts even if the user/listener isn't using them?  What sort of bandwidth cost is that going to be?  I assume we're either paying to load these from a CDN, or more likely the Nextjs app sending the fairly large JSON to the browser.

File: components/Player/Player.tsx
Comment (cavis) [inline components/Player/Player.tsx:394]: What's this about?  Seems like a no-op `if`.

File: components/Skeleton/Skeleton.tsx
Comment (cavis) [inline components/Skeleton/Skeleton.tsx:3]: So this is just used while the transcript is loading - and then replaced with the actual transcript element?

File: lib/convert/string/convertJsonToVtt.spec.ts
Comment (cavis) [inline lib/convert/string/convertJsonToVtt.spec.ts:51]: Okay, so here are some tests on that conversion process.  Cool!

File: lib/convert/string/convertSrtToVtt.ts
Comment (cavis) [inline lib/convert/string/convertSrtToVtt.ts:6]: Do these then get converted to JSON, which is what the browser is using?  (So we're always going `srt -> vtt -> json`)

**Review by cavis — APPROVED** [review]
Okay, skipped a bunch of boilerplate-looking changes and scss files.  But did have a few questions on some of the transcript parsing/converting logic.  And a couple other places.

Some optional DRY-refactors, but nothing that should really hold this up.  Overall, looks good to me 👍 .

File: lib/math/number/roundToN.ts
Comment (rpeterman-gp) [reply lib/math/number/roundToN.ts:3]: Yep. Javascript is fun. I tried the `Number.EPSILON` approach, but ran into issues with it. Ended up on this solution that seemed to work consistently.

File: pages/api/proxy/transcript/index.ts
Comment (rpeterman-gp) [reply pages/api/proxy/transcript/index.ts:64]: I had thought about this, but not all use cases were identical, at least at first. The checks all seem the same, but the processing function return values differ depending on the endpoint logic later on (eg. made sense to process to JSON object or just JSON string).

File: lib/parse/string/parseVttSpeaker.ts
Comment (rpeterman-gp) [reply lib/parse/string/parseVttSpeaker.ts:7]: I can add a link to speaker denotation the VTT standard docs.

File: pages/api/proxy/transcript/index.ts
Comment (rpeterman-gp) [reply pages/api/proxy/transcript/index.ts:104]: I just figured it out. This endpoint is tailored to what worked best to render transcripts in our app designs, and give us the flexibility to adapt the logic to accommodate changes later. Also was kind of necessary since there isn't a standard around combining cues that spans all the transcript data types this endpoint supports as input. Had to play it by ear to get the result that felt best.

File: components/Player/Player.tsx
Comment (rpeterman-gp) [reply components/Player/Player.tsx:51]: The only transcript data that gets loaded prior to the user opening the CC view is the the VTT, which is loaded by the browser when it encounters the track element. The VTT files have seemed to be fairly small, even for the longer Buzzcast transcript I have been working with. The choice to even use the track element was based on accessibility, since users may be using technologies that assist using the track data without having to use our CC feature.

Comment (rpeterman-gp) [reply components/Player/Player.tsx:51]: Larger transcript files, like the JSON format files, do not get loaded until the user opens the CC view.

File: components/Player/Player.tsx
Comment (rpeterman-gp) [reply components/Player/Player.tsx:394]: Oh, this is old code. Originally the player used an in-memory audio elements. Had to switch to rendering an audio element with track elements. Couldn't find a good way to add tracks to the in-memory audio element that was less complicated than just rendering a track element. This can be removed.

File: components/Skeleton/Skeleton.tsx
Comment (rpeterman-gp) [reply components/Skeleton/Skeleton.tsx:3]: Yes. It gives the transcript area something to show while it is loaded and rendered, which can take some time due to the size of the files being retrieved and the shear number of elements being rendered.

File: lib/convert/string/convertSrtToVtt.ts
Comment (rpeterman-gp) [reply lib/convert/string/convertSrtToVtt.ts:6]: The VTT endpoint stops at VTT. Browser track elements will use the VTT data. JSON endpoint will go to JSON. JSON data is used to provide the high-fidelity highlighting and are loaded separately only when CC view is opened or episode is viewed on listen page. The index endpoint goes to a custom data model to facilitate rendering requirements of the transcript on episode pages.

File: pages/api/proxy/transcript/vtt.ts
Comment (cavis) [reply pages/api/proxy/transcript/vtt.ts:70]: Ah, okay.  That makes sense as a last-fallback.

File: components/Player/Player.tsx
Comment (cavis) [reply components/Player/Player.tsx:51]: Cool, makes good sense.

File: lib/fetch/transcript/fetchAudioTranscriptData.ts
Comment (rpeterman-gp) [reply lib/fetch/transcript/fetchAudioTranscriptData.ts:11]: It seemed like a data point that would always be available and reflect a change in audio content (ie. addition/change in ads). Added the param just to attempt to ensure updated transcripts get loaded in cases were those file paths would not have change between ad changes.

**Review by farski — COMMENTED** [review]
For the remaining work/rework coming out of this PR, it'd be good to open some tickets to capture those things. I know there's talk of making some of the features in this PR user preferences; I think that adds a level of complexity that may go beyond the original scope of what this project was intended to do. My preference would be to get these new features in front of users, and see if there's a need to that sort of thing. There may be a clear enough preference that we can be fairly opinionated on this, to keep the app simple, both in terms of development and maintenance, and UX.

**Comment by rpeterman-gp** [comment]
@farski To be sure that the expected tickets are created accurately, can you create those tickets then give your final approval on this PR when you feel that remaining work has been covered for our future attention? I want to avoid any more delays going back-and-forth with the creation of those tickets.

**Review by farski — APPROVED** [review]
I will not have time to create any tickets before Friday

**Pull request merged**
